text/3371-cargo-target-base-dir.md

```diff
@@ -0,0 +1,402 @@
+
+### Providing backlinks
+
+Backlinks are metadata in templated `CARGO_TARGET_DIR` that links target directories back to the workspace they came from.
+
+[`targo`][tg] uses them in its own form of the feature and `cargo` uses them too for templated `CARGO_TARGET_DIR`.
+
+While details of the stored data are left to the implementation (there is no need for `cargo` to expose this data directly, though it could be exposed through `cargo metadata` in the future, see the relevant section below), one could imagine using it to clean target directories whose corresponding workspace does not exist anymore when calling something like `cargo clean --all-workspaces` (doing it automatically is not possible, else any workspace on external disks would have its target directory cleaned up each time the disk is unmounted, which is way too aggressive a default).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+<!-- Why should we *not* do this? -->
+
+## One more option to find the target directory
+
+This introduces one more option to look at to find the target directory, which may complicate the life of external tools.
+
+This is mitigated by the forward link provided by default by `cargo` when using the templated form of `CARGO_TARGET_DIR`.
+
+## Hitting windows path length limits
+
+Depending on what naming scheme is used (e.g., a very long hash), we could hit the Windows path length limits if not careful.
+
+A mitigation for this is recommending a short prefix (in `CARGO_TARGET_DIR`) and using a hash that doesn't include that many characters but those are only mitigations and do not fully fix the underlying problem.
+
+## Forward links
+
+There a few cases where a symlink instead of a real dir will break programs: at least SQLite 3 can be configured to raise an error if the database is behind a symlink anywhere in its opening path, it's probably other programs can also be configured to check this (or do it by default). Since templated `CARGO_TARGET_DIR` won't become a default in this RFC, we are not breaking any existing use cases.
+
+## Transition period
+
+During the transition period, any `CARGO_TARGET_DIR` that was defined as containing `{manifest-path-hash}` will change meaning. `cargo`, for at least one stable version of Rust, should provide warnings about this and point to either this RFC or its documentation to explain why the incompatiblity arised and how to fix it. In practice, paths with `{` or `}` in it are unlikely, even more with the exact key used by cargo here, so maybe no one will ever see the warning, but it's better than silently breaking workflows.
```

Comment:
We have competing demands
- Users will want to set this and not worry about their Rust version
- Versions of Rust prior up to this being implemented will do a very wrong thing with variables set

We likely need to error, instead of warn.  While this will be frustrating for users who have to work with versions that error, this will raise awareness so they are less likely to set this with versions from before anything was implemented.

Reply:
What if we instead treat the target dir as unset?  I'm unsure if that would be a positive or negative type of clever.

Reply:
I'm not in favor of this: silently treating the directory as unset would be surprising when one was explicitely (if wrongly) set. I regularly use this feature and I would prefer and error telling what to fix than my artifacts jumping places in some cases but not other

Reply:
The "what to fix" is to turn off the feature completely when running specific versions of Rust (while you want it for others).

I feel like this is a "no win" situation and we need to make sure we call out the alternatives on this.

Reply:
Done in I think https://github.com/rust-lang/rfcs/commit/a821d25a624e1ddf3143e854a3f77dc9cff07dd9